eval/adjust.rkt

```diff
@@ -71,333 +71,263 @@
                                            (vector-ref vstart-precs (- n varc)))
                                         (*base-tuning-precision*)))
-
-    ; This case is weird. if we have a cancellation in fma -> ival-mult in fma should be in higher precision
-    (when (equal? op ival-fma)
-      (set! final-parent-precision (+ final-parent-precision (car new-exponents))))
     
-    (when (>= final-parent-precision (*rival-max-precision*))         ; Early stopping
+    (when (>= final-parent-precision (*rival-max-precision*))       ; Early stopping
       (*sampling-iteration* (*rival-max-iterations*)))
+    
+    ; Final precision assignment
     (vector-set! vprecs-new (- n varc) (min final-parent-precision (*rival-max-precision*)))
 
+    ; Intro and ampl propogation for each tail instruction
     (for ([x (in-list tail-registers)]
-          [new-exp (in-list new-exponents)]
+          [ampl (in-list ampls)]
           #:when (>= x varc)) ; when tail register is not a variable
       ; check whether this op already has a precision that is higher
-      (when (> (+ exps-from-above new-exp) (vector-ref vprecs-new (- x varc)))
-        (vector-set! vprecs-new (- x varc) (+ exps-from-above new-exp))))))
+      (when (> (+ intro ampl) (vector-ref vprecs-new (- x varc)))
+        (vector-set! vprecs-new (- x varc) (+ intro ampl))))))
+
+(define (exp-is-inf? x)
+  (equal? x 9223372036854775805))
+(define (exp-is-0? x)
+  (equal? x -9223372036854775807))
+; this function imitates an exponent of infinity as 9223372036854775805
+(define (replace-inf x)
+  (if (equal? x -9223372036854775805)
+      (abs x)
+      x))
+(define (crosses-zero? x)
+  (not (equal? (mpfr-sign (ival-lo x)) (mpfr-sign (ival-hi x)))))
+
+; We assume that NaNs can not be at this step in the interval
+(define (maxlog x)
+  (define lo-exp (replace-inf (mpfr-exp (ival-lo x))))
+  (define hi-exp (replace-inf (mpfr-exp (ival-hi x))))
+  (cond
+    [(and (exp-is-inf? hi-exp) (exp-is-inf? lo-exp))  ; x = [-inf, inf]
+     (get-slack)]
+    [(exp-is-inf? hi-exp)                             ; x = [..., inf]
+     (+ (max lo-exp 0) (get-slack))]
+    [(exp-is-inf? lo-exp)
+     (+ (max hi-exp 0) (get-slack))]                  ; x = [-inf, ...]
+    [else
+     (+ (max lo-exp hi-exp) 1)]))                     ; x does not contain inf, safe with respect to 0.bf
+
+(define (minlog x)
+  (define lo-exp (replace-inf (mpfr-exp (ival-lo x))))
+  (define hi-exp (replace-inf (mpfr-exp (ival-hi x))))
+  (cond
+    [(exp-is-0? lo-exp)                               ; x = [0.bf, ...]
+     (- (min hi-exp 0) (get-slack))]
+    [(exp-is-0? hi-exp)                               ; x = [..., 0.bf]
+     (- (min lo-exp 0) (get-slack))]
+    [(crosses-zero? x)                                ; x = [-..., +...]
+     (- (min lo-exp hi-exp 0) (get-slack))]
+    [else
+     (- (min lo-exp hi-exp) 1)]))                     ; x does not contain zero, safe with respect to inf
```

Comment:
I don't think we need to subtract 1 here. The reason we add 1 in the `maxlog` case is because we're ignore the mantissa, which is in [1, 2]. Not an issue with `minlog`

Reply:
That's true, I'll remove it